strings update

lachlan-00 opened the pull request:
code i used to grep the translated strings
```
grep -R -oh -E "T_\(.*\)" > ../regex-strings.txt
```
[regex-strings.txt](https://github.com/ampache/ampache/files/3537658/regex-strings.txt)

@kuzi-moto This is ready for you to check case and look over any weirdness i might have made. especially around any %s / $s strings.

It would also be good if you could check these for me and see if there's a better way to do it or at least hints for them.
```
T_($access)
T_($action['title'])
T_($f_name)
T_($fsubcat)
T_($key)
T_($msgid)
T_($original, $plural, $value)
T_($pref['description'])
T_($preference['description'])
T_($preferences['title'])
T_($result['error'])
T_("%1$s %2$s (Playlist)")
T_("%1$s %2$s (Smart Playlist)")
```

Good news is this not only adds a lot of new strings but reduces the total by about 200 strings while making it more natural

lachlan-00 commented:
@Psy-Virus  i think this is time to have a look over finalising strings that have needed changes for a long time.

I don't know what your preferred order is but i'll be looking over all the strings this week and then we will need to update the base messages.pot and translatable-database-strings.txt for transifex.

I want to be able to do this once and not do it again before 4.0.0. I've already made a lot of text/case changes in the base code and i don't want to make it any more difficult by changing translatable text after this.

If you have any thoughts let me know while this is ongoing.

lachlan-00 commented:
@Psy-Virus I think i'm done here.

Ready to torture our translators with a pile of string changes.

Psy-Virus commented:
It'll take some time to look over it. But I'll give feedback as ASAP.

:)

lachlan-00 commented:
Thanks @Psy-Virus
I really don't like that i'm making so many string changes but it does reduce the total count of translatable strings while also unifying the style/theme so i think it's worth hassle.

It's at least worth doing this time and ideally not for many years again.

kuzi-moto commented:
Agreed. I think this will help improve Ampache's image and hopefully from now on can work on keeping additions consistent with the style.

lachlan-00 commented:
@Psy-Virus i'm starting to get conflicts now, so i'll be looking at merging this on Monday.

Over all i'm happy that the reduction in strings and consolidation is enough to justify the changes. I just need to know if anything in this pull will affect transifex and whether i need to remove/change anything in that before we do.

Psy-Virus commented on admin/access.php:
```diff
@@ -90,7 +90,7 @@
         $access->update($_POST);
         if (!AmpError::occurred()) {
-            show_confirmation(T_('Updated'), T_('Access List Entry updated'), AmpConfig::get('web_path') . '/admin/access.php');
+            show_confirmation(T_('Updated'), T_('Access List entry updated'), AmpConfig::get('web_path') . '/admin/access.php');
```
Why one time a long version with Ampache personally speaking to the user and one time not?

I'd either do one or the other like:

Your Access List entry has been removed
and
Your Access List entry has been updated

Or

Access List entry removed
and
Access List entry updated

The all in all question is: Should Ampache speak personal/natural to the user or more generic? ^^

lachlan-00 replied:
@kuzi-moto this is probably the last major thing we need to check before finalising this pull.
personal sentences to the user. especially where we are personal in one and impersonal in the other

Psy-Virus commented on admin/catalog.php:
```diff
@@ -107,7 +107,7 @@
         if ($deleted) {
             show_confirmation(T_('Deleted'), T_('Catalog and all associated records have been deleted'), $next_url);
```
The/Your Catalog and all associated records have been deleted?

Psy-Virus commented on admin/catalog.php:
```diff
@@ -125,12 +125,12 @@
+            $body = count($songs) . ' ' . nT_('Song Enabled', 'Songs Enabled', count($songs));
         } else {
-            $body = T_('No Disabled Songs selected');
+            $body = T_('No disabled songs selected');
```
Why writing "disabled songs" lower case when you mostly write the stuff, that's the current subject, uppercase?

Psy-Virus commented on admin/catalog.php:
```diff
@@ -125,12 +125,12 @@
                 Song::update_enabled(true, $song_id);
             }
-            $body = count($songs) . nT_(' Song Enabled', ' Songs Enabled', count($songs));
+            $body = count($songs) . ' ' . nT_('Song Enabled', 'Songs Enabled', count($songs));
```
Good one, I like that... should definitely have been done years ago :D

Psy-Virus commented on admin/catalog.php:
```diff
@@ -233,7 +233,7 @@
             require AmpConfig::get('prefix') . UI::find_template('show_disabled_songs.inc.php');
         } else {
-            echo "<div class=\"error\" align=\"center\">" . T_('No Disabled songs found') . "</div>";
+            echo "<div class=\"error\" align=\"center\">" . T_('No disabled songs found') . "</div>";
```
Here again; subject: "Disabled Songs"

Psy-Virus commented on admin/catalog.php:
```diff
@@ -125,12 +125,12 @@
         }
         $url      = AmpConfig::get('web_path') . '/admin/catalog.php';
-        $title    = count($songs) . nT_(' Disabled Song Processed', ' Disabled Songs Processed', count($songs));
+        $title    = count($songs) . ' ' . nT_('Disabled Song Processed', 'Disabled Songs Processed', count($songs));
```
"Processed" should probably be lowercase

lachlan-00 replied:
agree

Psy-Virus commented on admin/mail.php:
```diff
@@ -64,10 +64,10 @@
             } else {
                 $title     = T_('E-mail Not Sent');
-                $body      = T_('Your E-mail was not sent.');
+                $body      = T_('Your e-mail was not sent.');
```
"was" vs. "hasn't been" vs. "has not been"

Psy-Virus commented on admin/mail.php:
```diff
@@ -64,10 +64,10 @@
-                $body   = T_('Your E-mail was successfully sent.');
+                $body   = T_('Your e-mail was successfully sent.');
             } else {
                 $title     = T_('E-mail Not Sent');
```
"sent"?

lachlan-00 replied:
this is returning on success so this part is sent. (rewriting this)

Psy-Virus commented:
A few years ago, I also thought about making Ampaches strings and "speech" more natural and consistent. But then I realized that this is just a crazy amount of work XD

And sorry for the late response but I had some really exhausting days.

kuzi-moto commented:
@Psy-Virus The majority of the changes that I proposed were mostly just capitalization, and then the rest small grammar fixes. I agree that changing to a more consistent style between all strings would be a good thing to do and would definitely help in that endeavor. Personally, I'm for more concise messages and don't mind if they aren't more like natural speach, but if the consensus is more toward the latter than that is fine with me.

With that being said, we also need to establish some rules around capitalization. You mentioned that for `No disabled songs selected` you should capitalize "disabled songs" because it's the subject. I would only argue that there are no grammar rules for capitalizing the subject of a sentence. "Disabled" is just an adjective that describes the noun "songs", but is "disabled songs" considered a pronoun in Ampache? Generally only the first word and pronouns are capitalized when using sentence case. If something is a title, I think all words should use title case. In looking through Ampache's strings, it seems there are a few functions being used that can have some standardization for all strings used in them:

* class UI: `show_box_top($title = '', ...)` - $title should probably be title case.
* class AmpError: `add($name, $message, ...)` - $message probably sentence case. Also, for this one, a standard format for errors? Sometimes the message is "Error: ..." and other times without.
* ui.lib.php: `show_confirmation($title, $text, ...)` - It seems like $title should mostly be title case, $text probably always sentence case. We should standardize this though probably as $title is in title and sentence case.
* Some strings are assigned directly to variables like $title and $body, title and sentence case should be used respectively.
* Some text is put directly into HTML code, such as div's. These would depend on the context they are being used, but most likely in sentence case.
* Some is just echoed out in the code.

There's a little over 3,200 instances of `T_(` in Ampache code. I'm sure if we can come to an agreement on how to proceed, it would not take long to review since there are a lot of duplicates.

By the way, I'd like to propose changing e-mail to email. 6,760,000,000 results for "e-mail", 15,420,000,000 for "email".

Thanks again for looking it over. I know all the contributors of Ampache have put a lot of time and effort for no pay, and for it to be where it is today is amazing!

lachlan-00 commented:
Okay, good start.
I don't need a full review and waiting for weekends aren't a big deal to wait for me. I just wanted to have some different perspective.

Next step then is to look at the context of strings and consistency.

lachlan-00 commented:
I'm not sure of my preference for personal or impersonal speech but it definitely has to be one

Psy-Virus commented:
> With that being said, we also need to establish some rules around capitalization. You mentioned that for No disabled songs selected you should capitalize "disabled songs" because it's the subject. I would only argue that there are no grammar rules for capitalizing the subject of a sentence. "Disabled" is just an adjective that describes the noun "songs", but is "disabled songs" considered a pronoun in Ampache?

My intention wasn't to develop new grammar rules XD
I was mentioning it just because if you take a look at most of the strings, then you'll see that the subject, the string is describing, is mostly capitalized.

> If something is a title, I think all words should use title case. In looking through Ampache's strings, it seems there are a few functions being used that can have some standardization for all strings used in them:

Sounds good

> By the way, I'd like to propose changing e-mail to email. 6,760,000,000 results for "e-mail", 15,420,000,000 for "email".

Well, that's a way do decide it. But; the strings used in Ampache are defined as US-English. We should consider that too if we change something. For example the word "catalog"
EN-US = Catalog
EN-GB = Catalogue

After some research I came to the conclusion that there's no defined rule for e(-)mail but American publications use e-mail more than email.
But personally I'm also fine if we let Google decide it :D


> + class UI: show_box_top($title = '', ...) - $title should probably be title case.
> + class AmpError: add($name, $message, ...) - $message probably sentence case. Also, for this one, a standard format for errors? Sometimes the message is "Error: ..." and other times without.
> + ui.lib.php: show_confirmation($title, $text, ...) - It seems like $title should mostly be title case, $text probably always sentence case. We should standardize this though probably as $title is in title and sentence case.
> + Some strings are assigned directly to variables like $title and $body, title and sentence case should be used respectively.
> + Some text is put directly into HTML code, such as div's. These would depend on the context they are being used, but most likely in sentence case.
> + Some is just echoed out in the code.
>
> There's a little over 3,200 instances of T_( in Ampache code. I'm sure if we can come to an agreement on how to proceed, it would not take long to review since there are a lot of duplicates.

I'm happy with this

> I'm for more concise messages and don't mind if they aren't more like natural speach

I like the natural speech more. I like it if programs speak personal and natural to you.
Like for example:

+ The song has been deleted
vs.
+ I deleted this song (for you)

Or if Ampache should do something:

+ Disable
vs.
+ Disable it

I just think that this gives a more personal feeling.
Ofc those are not the best examples but I guess u get what's in my mind.

:)

Psy-Virus commented:
Oh and there's something else:

If we have something not string related within our strings that should be translated; like HTML or links we should avoid to put them in. I know there are still some strings that contain HTML or links.

In those cases we should ALWAYS search for a way to exclude the code or link or string foreign elements.
For example with: https://www.php.net/manual/en/function.sprintf.php

Otherwise translators have to include this in the translated texts what could lead to mistakes.

Also; gettext supports Translator Hints.
I already did this for many strings containing links and stuff.
e.g.

`header_inc.php`

```php
printf(json_encode(nl2br(
    /* HINT: Translator, "%s" is replaced by "cookie settings" */
    T_("We have placed cookies on your computer to help make this website better. You can change your %s at any time.\nOtherwise, we will assume you are OK to continue.\n\nClick on this message to not display it again."))),
    "<a href=\"" . AmpConfig::get('web_path') . "/cookie_disclaimer.php\">" . T_('cookie settings') . "</a>"
);
```

This hint is showing in Transifex like the stuff in the red square:

![grafik](https://user-images.githubusercontent.com/6222200/61578002-fa9cda80-aaef-11e9-9cf6-82897fcc0186.png)

lachlan-00 commented:
I made a choice to go with personal text and it seems to be much nicer to read.

Lot's of sprintf additions which will cut down on a lot of duplications.

In en_US english e-mail is the standard for print text. but is not common outside of the US. (I'm not american but i feel like this follows the rules.)

Psy-Virus commented:
Hey @lachlan-00,
what exactly is the reason to use sprintf in the following example?

```php
        $title = sprint_f(T_('Do you really want to disable this %s?'), T_('module'));
```

I think 
```php
        $title = T_('Do you really want to disable this module?);
```
is the better way to go, as each extra function you're not calling makes the code faster.

Also, you're two times calling gettext instead of one time. This theoretically doubles the processing time as the string length doesn't really matter in this case.

Psy-Virus commented:
> I made a choice to go with personal text and it seems to be much nicer to read.

I like the changes.

> In en_US english e-mail is the standard for print text. but is not common outside of the US. (I'm not american but i feel like this follows the rules.)

So we stick with "e-mail" in the code base?

kuzi-moto commented:
> what exactly is the reason to use sprintf in the following example?
>
>         $title = sprint_f(T_('Do you really want to disable this %s?'), T_('module'));
>
> I think
>
>         $title = T_('Do you really want to disable this module?);

I thought this was to separate common words, so like "module"  wouldn't have to be translated many times?

> So we stick with "e-mail" in the code base?

Yeah, we probably don't need to change it.

@lachlan-00 Any comments on my previous post? I'd like to help with strings, but don't want to start without finalizing how we want to deal with everything.

lachlan-00 commented:
@kuzi-moto I'm not too set on how deep our rules should be but I think we look at some basic rules. keep it simple as possible.

* All comments and code are to written in US english.
* Ampache Pages / features / Items are capitalised in all sections. (Show Now Playing for user, Edit Playlists, Look at these Albums)
* Header lines / Titles are capitalised. (You Mission Was Successful)
* Informational / body text is sentence case. (The album you asked me to delete was removed.)

@Psy-Virus I split these out to reduce the amount of translatable strings while making more sections of text uniform but it depends on what direction we want to go.

Do we shoot for more full sentences that make translating easier but create more things to translate?
I'm not against pulling that back, as the strings themselves have been unified anyway.

I think one issue may be the structure of sentences in english doesn't mean these strings work in other languages which might make full sentences more resilient in other languages.

kuzi-moto commented:
Sounds good! I'll compile a list of Ampache things that should be capitalized, and maybe make a wiki page about it for others.

Something I thought about, what if we edit Amperror to append "Error:" before each error, so you don't have to specify it in each message. Something like this:

```PHP
 public static function add($name, $message, $clobber = 0)
    {
$message = T_('Error') . ': ' . $message
...
```

Sudocode because I don't know much PHP offhand, but then error will be appended to each error message, and won't have to be included in so many strings. I think it would make for easier translation?

lachlan-00 commented:
There might be ways to make those into a more casual sentence without putting a prefix in.

When the title of messages States there was a problem:
"Please don't be mad" / "I'm sorry but this didn't work", etc
The body doesn't need to remind you it's an error; just tell you what happened

**I'm sorry**
I couldn't remove this file

**I had a problem**
The album does not exist

lachlan-00 commented:
for the non-popup messages that are one line there should be a cleaner way to report an error without the formal "Error: %s"

Even just something like "Sorry, that file can't be deleted."

Psy-Virus commented:
> @Psy-Virus I split these out to reduce the amount of translatable strings while making more sections of text uniform but it depends on what direction we want to go.

This will definitely happen. Already happened before^^
Only strings with foreign elements in them, like HTML should be handled via s_printf.

In this example 

If we have something like this:

```php
$title = sprint_f(T_('Do you really want to disable this %s?'), T_('module'));
```

The translator doesn't see right away that a module is meant.
Module for example is a neutral noun in English and German for example.
In German we have to put the "das" or "dieses" in front of it.
We have 3 different Articles. The male, female and neutral (Der, Die, Das) I think French has the same amount.

So the complete sentence would be one of the following: 
"Möchtest/Willst Du __das Modul__ wirklich deaktivieren?"
"Bist Du Dir sicher, dass du __das Modul__ deaktivieren möchtest/willst?"

 But if I'm in Transifex, I don't know with what %s will be replaced (Exept there's a Translator Hint as described in [my Post](https://github.com/ampache/ampache/pull/1957#issuecomment-513458743) somewhere ⬆️).

kuzi-moto commented:
Are you saying we would have to put a note in the code every time that sprint_f is used so translators know what will be there?

Psy-Virus commented:
> Are you saying we would have to put a note in the code every time that sprint_f is used so translators know what will be there?

That would be the only way, the translator sees right away how he/she can build the sentence to integrate the variable/string replacement.

lachlan-00 commented:
Sounds like it's better to pull them back to sprintf's with hardcoded string replacement then.

lachlan-00 commented:
@kuzi-moto i tried to make a list of the current strings from the update today.

This should hopefully help in cleaning up. After all this work we've cut down on over 200 superfluous strings.
[amount.txt](https://github.com/ampache/ampache/files/3446358/amount.txt)

kuzi-moto commented:
Ugh, I was looking through the strings... It looks like Ampache uses "catagory" a ton, when it should be "category"... Obviously this is going to take a huge effort to fix as it's used in a lot of areas. Before I go and make any changes, would it be best to just change it where it will actually show up in the interface, and not in the backend?

Psy-Virus commented:
> Ugh, I was looking through the strings and it looks like Ampache uses "catagory" a ton, when it should be "category"... Obviously this is going to take a huge effort to fix as it's used in a lot of areas.

Notepad++ / Atom -> search within files -> replace "this" through "that" -> Pull Request Changes.

kuzi-moto commented:
I'd be okay with doing that for just strings/plaintext, but what about SQL tables? there is "catagory", and "subcatagory" in the preference table, as well as in the "label" table.

lachlan-00 commented:
Leave the table names and preference names in the code. it's not worth changing db names for the sake of it.

It's only the text that a user sees we need to worry about and outside of those I don't really see that many?

lachlan-00 commented:
@kuzi-moto in your pull another thing to remember is that the ampache.sql file has a minimum version much higher than all the updates in update.class.
```
INSERT INTO `update_info` (`key`, `value`) VALUES
('db_version', '380004'),
```
You could actually just delete the updates <= 380004 and there won't be any issues.

kuzi-moto commented:
A couple more questions as I was browsing through code.

* There are some sections that are not translated, should they be?
* When a variable is at the beginning or end of string, would it be better to remove it from the T_ function?
* In some cases with the show_confirmation function it's got a question as both title and body. Would it make more sense to combine the phrases into one? It just sounds weird to me. Ex:

```php
show_confirmation(T_('Are you sure?'), T_('Do you really want to delete this Album?'),
```

Maybe instead something like

```php
show_confirmation(T_('Are you sure you really want to delete this Album?'),
```

lachlan-00 commented:
show_confirmation is the visual notifications inside the page

![image](https://user-images.githubusercontent.com/1305249/62184378-2a29cd80-b3a1-11e9-98e0-dc904b93f3a0.png)

I'm note really sure if it's a great idea to use a single line for those as the text for some can be long.

kuzi-moto commented:
> I'm not really sure if it's a great idea to use a single line for those as the text for some can be long.

There's a bunch that are single lines. You think they should be updated use both lines?

lachlan-00 commented:
I'm sure some would be fine with a body / title only but I think you'd have to make a judgement whether it looks right on the page.

kuzi-moto commented:
Yeah, that makes sense. Any thoughts on my other bullets?

lachlan-00 commented:
There have to be a few strings that should be translated. It might be easiest to change language and see if you can see anything not in English or at least strings that exist in translations but aren't in a T_() function.

It would be good to remove superfluous symbols like ' - ', etc but it might not be easiest. maybe a hint that they are cli or the location?

Psy-Virus commented:
> I'd be okay with doing that for just strings/plaintext, but what about SQL tables? there is "catagory", and "subcatagory" in the preference table, as well as in the "label" table.

We definitely have to rework this somehow. Those strings stored in the DB are kinda cancer.

> There are some sections that are not translated, should they be?

which, where?

> When a variable is at the beginning or end of string, would it be better to remove it from the T_ function?

Do you have an example?

> Same goes for symbols. Should we try to remove symbols that aren't punctuation? Such as ellipsis, or colons at the end of a string?

But what to do if a word or sentence is shortened and can be fully visible by hovering over it?
if it's a full string we should remove it. Except it's telling you that something in running in the background and you have to wait.

> I'm sure some would be fine with a body / title only but I think you'd have to make a judgement whether it looks right on the page.

This or that way, we should keep a consistent way to show messages/errors/infos

kuzi-moto commented:
> > When a variable is at the beginning or end of string, would it be better to remove it from the T_ function?
>
> Do you have an example?

`printf(T_('Reading Catalog: %s'), $catalog->name);` I think lachlan put in a hint for this one, but I'm not sure it's necessary to include?

kuzi-moto commented:
Hey @lachlan-00, how close do you think we are to merging this?

lachlan-00 commented:
This and subsonic are the 2 big ones on the list to complete so i need to sit down on this branch again and go over it. I'd like it to be soon but i also don't want to release it to translation without being sure it's all done in a better way.

lachlan-00 commented:
@kuzi-moto i'm going to do another run looking for echos/prints without translation. I've been through T_( regexes so many times now i'm not sure there is much left.
One thing we also need to check is where there are strings which i just variables. These should have hints or be edited to show some kind of context.

I know there are a few for the ACL/Access names but i'm sure there are more.

lachlan-00 commented:
Just looking now and it's somewhere between 500-1500 echo's to look at, just cleaning up my search.

lachlan-00 commented:
@kuzi-moto  and @Psy-Virus  one last thought for both of you. I think this will dictate the remaining changes that need to happen.

Is ampache going to talk to the user as an individual?
"I'm sorry"
"I have deleted this playlist"
"I've sent you an e-mail with your login details."

Or as an abstract person?
"Don't be mad"
"Your playlist has been deleted"
"Please check your e-mail for login details"

I think i've split this both ways so i'm not actually sure which i prefer. I lean towards only talking to you as an abstract person and avoiding I/me

kuzi-moto commented:
I have thought about this as well, and what would sound better. I think I prefer being more abstract though. There are probably some cases where Ampache talking in first person could be neat, but overall would be better as more general.

lachlan-00 commented:
Okay, I'm going to look at transifex with some beers this weekend and try to finish this off.

* Abstract, natural language.
* Hints for all variables
* Try to put variables at the end of lines
* No random punctuation in strings.
* Change my language to French and check what's missing.
* Ping kuzi to check case and to have someone else read changes.
* Ping psy to do another review.

The difference in the strings is a world better than before but I want to not make such a major change again after this

kuzi-moto commented:
Sounds like a plan! It's definitely not a good idea to massively change all the text more than once... But there was certainly a lot that really needed to be fixed.

Psy-Virus commented on bin/install/install_db.inc:
```diff
@@ -108,25 +108,26 @@ if (!install_create_config()) {
+    echo "install_db.inc [" . T_('Options') . "]";
     echo "\n\t-U, --database-user\t\t\t";
-    echo 'MySQL Admin User';
+    echo T('MySQL Administrative Username');
```
Is there a reason for the missing underscores?

lachlan-00 replied:
No! my wireless keyboard strikes again.

Psy-Virus commented:
Well, there are a lot of changes. But it looks really good.
I didn't see any problems yet. But to be honest, after looking through hundreds of lines there might be something, I missed :D
But this will be something we obviously can fix later.

Did one of you maybe look into the modules? And check how we can get around the database strings?

lachlan-00 commented:
The only way i think we can resolve hidden strings like the db modules is by hardcoding their id numbers.

translatable-database-strings.txt has table id numbers which are fine for the default installs but the modules will change number when added/removed so i think that would be the only way around it. which i don't think is easy either.

lachlan-00 commented:
i'm going to vietnam on saturday for a full week so lets sit on this till i get back. (9/10 sep depending on your timezone.)

The subsonic merge will have a chance to settle then and @Psy-Virus can have a week to think about the impending transifex apocalypse with all these changes. ;)

Psy-Virus commented:
> translatable-database-strings.txt has table id numbers which are fine for the default installs but the modules will change number when added/removed so i think that would be the only way around it. which i don't think is easy either.

The IDs in the `translatable-database-strings.txt` were/are only a reference for me and potential other translation admins/maintainers and have nothing to do with the translations itself. :)
The problem is just that the preference descriptions string thingies are added to the preference table when they're enabled and dropped when they're disabled.
And since the translation script scans the preference for strings to translate and puts them at the end of the `messages.po`, only active module strings are gathered.
So the strings are the problem, not the IDs.

A while back, I thought about a workaround and simply put all the strings into a file or right away into the preference table and determine if the string should be sown (module/plugin enabled) by having another boolean field in the preference table per string. But this could considerably slow down the process of echoing the strings and would be too complicated for something that simple.

I think the whole plugin/module system has to be rewritten in a Gettext/translation friendly way.
Maybe so that in the database is only stored, if the plugin or module is active and the translatable strings are just put into the corresponding files like it is for the rest of Ampache.

> i'm going to vietnam on saturday for a full week so lets sit on this till i get back. (9/10 sep depending on your timezone.)

Sounds great, hope you have a lot of fun!

> The subsonic merge will have a chance to settle then and @Psy-Virus can have a week to think about the impending transifex apocalypse with all these changes. ;)

Yepp, that's why I didn't provide a translation update for so long.
I mean I rewrote the German translation completely within 6 - 8 hours in a single run, fixed a bunch of stuff to make Ampache more friendly for foreign languages and maintain it since then.
The translations at all were catastrophic before. The only language that was more or less okay was the source language. So this should, depending on how many brainlags occurred while coding, and easier task.

:D

lachlan-00 commented:
I like the question first. so i'll change that now.
```
Are you sure?
If you continue your album will be deleted.
```

For admin levels i do think a reference to the int in some way makes a lot more sense as we use it  based on a 0-100 scale.

That might mean we change it to refer to the number first
```
T_('100 (Administrator)')
```
or as a reference in the translation.
```
T_('Administrator (100)')
```
Change them to denote that percentage value makes more sense to me rather than working out how admin vs catalog manager is different. I'll look at the access function now as well.

Regarding plugin strings i vote for moving their translation out of the database into the plugin file. That way we can keep their name and value for searching and then translate them from the file.

lachlan-00 commented:
@Psy-Virus i think i have the solution.
in 39ce341 i have translated the description for all /modules/plugins Preferences::insert.
This doesn't change enabled plugins but when they are reactivated they will store the translated name in the database.

This also doesn't need any other modifications as the name of the preference doesn't change. just the description.

I think the only question left now is will these files be picked up when generating the translations?

lachlan-00 commented:
@kuzi-moto if you want to make a choice about how we list the admin users and put that in a pull i think we're done

lachlan-00 commented:
Okay, I'm not touching this now outside of you both making any comments/pulls regarding the contents.

I think we've made the right choices here and overall have made a much more positive experience for text/translation. really excited to get this one in! :)

lachlan-00 commented:
make a choice on dropping the log out from the top bar too
![image](https://user-images.githubusercontent.com/1305249/63929995-c83eaf80-ca95-11e9-8a0c-351a7ba82732.png)

i think drop it

lachlan-00 commented:
also, why aren't we putting the playlists in the playlist section?
![image](https://user-images.githubusercontent.com/1305249/63931699-21f4a900-ca99-11e9-9ff8-5128a1b351a1.png)

lachlan-00 commented:
one last thing before i go.

does the order of the groups in the sidebar need changing? i find myself scrolling down to search more than using random.
![image](https://user-images.githubusercontent.com/1305249/63932026-bbbc5600-ca99-11e9-9d74-6e618c5b3809.png)

lachlan-00 commented:
I promise last thought before i leave this for everyone else to go over.

![image](https://user-images.githubusercontent.com/1305249/63945906-89b9ec80-cab7-11e9-997e-2a5e976d7738.png)

This filters stuff but what does it actually filter again? toggling artwork seems like it should be somewhere else as well. i keep forgetting this is there because it doesn't consistently show up for me.

kuzi-moto commented:
> make a choice on dropping the log out from the top bar too

Agree, seems a bit redundant to have log out next to username as well as in the 'tool panel'

> also, why aren't we putting the playlists in the playlist section?

Also agree, seems to make sense to move "Playlists" and "Smart Playlists" to the Playlists section.

> does the order of the groups in the sidebar need changing? i find myself scrolling down to search more than using random.

I think that would make sense. I vote we move "information" down to the bottom, and maybe "search" under "playlist"? To add to this, I feel like maybe it would be better to remove "Random" entirely, and just link the "Advanced" page under "Music" labeled as "Random"? Actually I feel the same way about "Search" since there is already the search bar at the top, unless that gets a lot of use by others, I just don't. Unfortunately I don't like the look of either the Search or Random pages, and I think need some work.

> This filters stuff but what does it actually filter again? toggling artwork seems like it should be somewhere else as well. i keep forgetting this is there because it doesn't consistently show up for me.

I was looking into this a while ago, because it really annoyed me. I believe it's using the context of the page it's on to appear. So if you're on the "Songs" page and you enter text in the filter, then it will show only songs that start with your query. It also works for filtering songs on an album page, filtering albums on the album page, etc. 

However it's really annoying because it will show up in some cases that don't really make sense, or just aren't useful. I think we should remove it from the sidebar, and instead add it individually to pages where it makes sense.

stebe commented:
Strange, I don't have a logout button. I remember I used to.  Right now the text log off is all I have.

![image](https://user-images.githubusercontent.com/4076671/63979551-23b57100-ca87-11e9-87f9-3158eeccc3be.png)

stebe commented:
> one last thing before i go.
> 
> does the order of the groups in the sidebar need changing? i find myself scrolling down to search more than using random.
> ![image](https://user-images.githubusercontent.com/1305249/63932026-bbbc5600-ca99-11e9-9d74-6e618c5b3809.png)

This is really a personal preference and ideally the sections layout should be user configurable. I rarely search from the sidebar and use the search on the top menu.  I like the results as I type feature. I use the sidebar for random because I like to be surprised. The search and random sections would be better at the very top The information section could be moved down. I've find no need for the Song, Albums or Artists as I have so much music that browsing isn't useful. Playlist import probably isn't used much so might be better on another screen. My opinion the section order would be Search. Playlists, Random, Music, Information.

stebe commented:
> I promise last thought before i leave this for everyone else to go over.
> 

> 
> This filters stuff but what does it actually filter again? toggling artwork seems like it should be somewhere else as well. i keep forgetting this is there because it doesn't consistently show up for me.

I never see the toggle artwork button. Maybe again from a lack of space to display this. Also the filters section doesn't disappear when I got back to the currently playing screen.

kuzi-moto commented:
> Strange, I don't have a logout button. I remember I used to. Right now the text Log Out is all I have. The difference is have a plugins button, I wonder if there isn't enough room for all the buttons.
> 
> ![image](https://user-images.githubusercontent.com/4076671/63979551-23b57100-ca87-11e9-87f9-3158eeccc3be.png)

Actually same for me. I would probably prefer the icon though.

lachlan-00 commented:
> 
> 
> > Strange, I don't have a logout button. I remember I used to. Right now the text Log Out is all I have. The difference is have a plugins button, I wonder if there isn't enough room for all the buttons.
> > ![image](https://user-images.githubusercontent.com/4076671/63979551-23b57100-ca87-11e9-87f9-3158eeccc3be.png)
> 
> Actually same for me. I would probably prefer the icon though.

The logout button is in the stringsupdate branch but not the develop branch.

Developo has the plugins button but stringsupdate has moved that into settings.

stebe commented:
The collapsed menu will need a Logout button.

![image](https://user-images.githubusercontent.com/4076671/63982262-df7a9e80-ca8f-11e9-912a-b3d3cbf38c4b.png)

lachlan-00 commented:
You're right @stebe the compact menu would need an option or a change to allow logout.

i think you're also right about moving import. it's more of a setting rather than a music item. maybe into catalog?

lachlan-00 commented:
It depends on the security level required but import could probably be in two places that are better than where it is now.

or possibly user tools in settings?
![image](https://user-images.githubusercontent.com/1305249/63982530-2d31ee00-cb06-11e9-9468-deba620a7fa1.png)

or in the user preferences?
![image](https://user-images.githubusercontent.com/1305249/63982569-4dfa4380-cb06-11e9-83e7-6793bc0f8bde.png)

lachlan-00 commented:
Merged in some of the suggestions from @stebe 

Reorder of home
![image](https://user-images.githubusercontent.com/1305249/63992974-ba3e6c80-cb31-11e9-95c4-c393f036cd2c.png)

Move playlist import into preferences tab
![image](https://user-images.githubusercontent.com/1305249/63992984-c62a2e80-cb31-11e9-9a21-556e5ad366c4.png)

user avatar in the edit menu (makes the delete button more sane as well)
![image](https://user-images.githubusercontent.com/1305249/63993005-e4902a00-cb31-11e9-8bce-fb9f9ffd2ec5.png)

added a check on the avatar upload which was silently dropping the image and not telling the user.
@kuzi-moto i made it tell me the dimensions but it's probably better to drop. (making more work for you!)
![image](https://user-images.githubusercontent.com/1305249/63993033-0689ac80-cb32-11e9-9b6d-03c97d1f8a5a.png)

I made a logout button for the light sidebar but i need to get back onto my linux box so i can gimp up the transparency
![image](https://user-images.githubusercontent.com/1305249/63993073-30db6a00-cb32-11e9-9081-da08a2390afa.png)

lachlan-00 commented:
please someone fix my terrible art or a CC png that could replace it.
![image](https://user-images.githubusercontent.com/1305249/63994976-f8d82500-cb39-11e9-92d5-e19fd0392cd1.png)

lachlan-00 commented:
On Holidays now.
@kuzi-moto and @stebe feel free to discuss/extend however you want
@Psy-Virus push their change in if there's anything extra while I'm gone.

Enjoy your weeks, i know i will

stebe commented:
> It depends on the security level required but import could probably be in two places that are better than where it is now.
> 
> or possibly user tools in settings?

.
> 
> or in the user preferences?

Importing a playlist is really a tool so would be more logical in User Tools.

stebe commented:
Deleted previous comment. My end didn't update. Looking at it now, it looks good.

lachlan-00 commented:
Interesting bug. Log out using the light menu on mobile... Same link but it looks weird and loses the logo for me.

But I'm glad you like the changes so far stebe. I'm excited to get this out there soon for everyone else.

kuzi-moto commented:
I rarely use any of the categories under "Music" so I don't have much preference there, but as @stebe mentioned, it would be nice if we could allow people to re-order them in the future. 

Thinking about it, someday it would be nice to implement anonymous metrics into Ampache, so we could tell which features and options are actually being used the most.

kuzi-moto commented:
Submitted my final major revision, went through all translatable strings and found a lot of things. After it gets approved, I'd like to test drive it for a few days/week to see if anything stands out.

Additionally, I wonder if we can consolidate and make all text around working with files the same, like there are a couple variations of "%s file was deleted", or "%s file was read", etc.  Not a big deal but would reduce some strings.

kuzi-moto commented:
Also, I was able to take care of some more of the variables mentioned in the OP. I think any relating to preferences will probably need to implement Lachlan's idea or something.

lachlan-00 commented:
I've been using this branch day to day before I left but the only way to make sure is to drive it more.

If psy doesn't want to merge your pull I'll do it when I get back and then fix a few of those new bugs before going to this again in a week or two.

The light menu logout needs to be fixed as well but that's the only bug this branch has

kuzi-moto commented:
Sounds like a plan! On a side note, and this isn't really that important as it serves no actual function, but what is `<!-- Propulsed by Ampache | ampache.org -->` and what does it even mean? I'm guessing it's some sort of typo because the only definition I can find for "Propulsed" is "to repel; to drive off or away" which doesn't make any sense in this context. I think we should change it to whatever makes sense.

Next, I wanted to ask about this:

```PHP
echo "<dt class=\"" . $rowparity . "\">" . T_($key) . "</dt><dd class=\"" . $rowparity . "\">" . $value . "</dd>";
```

I can't make any sense of how this is working. Here's a more complete example of one of the cases where it shows up:

```PHP
<?php
    $itemprops[gettext_noop('Name')]     = $radio->f_name;
    $itemprops[gettext_noop('Website')]  = scrub_out($radio->site_url);
    $itemprops[gettext_noop('Stream')]   = $radio->f_url_link;
    $itemprops[gettext_noop('Codec')]    = scrub_out($video->codec);

    foreach ($itemprops as $key => $value) {
        if (trim($value)) {
            $rowparity = UI::flip_class();
            echo "<dt class=\"" . $rowparity . "\">" . T_($key) . "</dt><dd class=\"" . $rowparity . "\">" . $value . "</dd>";
        }
    }
?>
```

So, it looks like the `$key` value is coming from `gettext_noop('Name')`, `gettext_noop('Website')`, etc. Looking at the `gettext_noop()` function, this is all it does...

```PHP
function gettext_noop($string)
{
    return $string;
}
```

What is the point of this? Can we just replace `gettext_noop()` with `T_()` and then change `T_($key)` to `$key` as it will already be translated?

lachlan-00 commented:
gettext_noop does seem totally useless. swapping with T_.

I'll test that out now. it was in gather-messages.sh for collecting messages but i think that's useless as well over T_

lachlan-00 commented:
Propelled by Ampache!

kuzi-moto commented:
Yes, nice! That was bothering me way more than it should have...

lachlan-00 commented:
i'll work on the logout bug then redo the strings regex and we might be close to the final stages!

kuzi-moto commented:
I agree, seems like it would make sense to not have to duplicate calls to T_ as much as possible. Seems like it's going to be a bit of a challenge as you mentioned, so that strings aren't being translated when they don't need to be. 

One thing I'd like to suggest, is how about storing the strings in an array? It might make it easier to read the code when you know a value being used is a translated string.

```PHP
$s = array(
    'home' => T_('Home'),
    'play' => T_('Play'),
    'artists' => T_('Artists'),
    'albums' => T_('Albums'),
    'playlists' => T_('Playlists'),
    'tagcloud' => T_('Tag Cloud'),
    'favorites' => T_('Favorites'),
    'upload' => T_('Upload'),
    'logout' => T_('Log out')
);

<span><?php echo $s['home']; ?></span>
```

I don't know how much use it would be, up to you!

lachlan-00 commented:
Having a string array might be useful for searching. I just made them single variables in templates for simplicity. Even a standard prefix would be better.

I think I'll just do the templates then that's enough. They're the things being constantly reloaded

kuzi-moto commented:
Makes sense! A prefix would be good too, something like $str_text might be even better as there's less syntax clutter.

lachlan-00 commented:
Let's do lowercase t. t_string

I've been grepping translations with it so let's make the variables similar

kuzi-moto commented:
Sounds good to me!

lachlan-00 commented:
okay @kuzi-moto i'm done with templates. I've moved the query stats to the bottom of pages when loaded so ajax loading will actually populate relevant data.

If you want to have a final go over when you're happy, we can ping psy for a final review and stage the changes when he will be ready to update translation files.

lachlan-00 commented:
So @kuzi-moto do you think there is anything left you need to look at here? i'm pretty happy with the interface as a whole now.

lachlan-00 commented:
Okay, @kuzi-moto lets move theme changes into new-theme now and other fixes into develop after this merges. I think if there are any strings left untranslated they will be dupes of existing translations that can be safely ignored for now.

@Psy-Virus it's your turn to decide when/how this should be merged. I'm happy for you to merge when you have the time to manage transifex updates.

I don't think this branch will change now so you could update transifex first? that might give translators time to read things over?

kuzi-moto commented:
@lachlan-00 I think I'm pretty good for now. I'm sure small stuff will come up every so often, but as you said safe to ignore. I've been using it a bit and haven't noticed anything bad.

lachlan-00 commented:
Final change here. Updated all plugins to translate the description.

This is just copy/paste so @kuzi-moto just double check the plugins description
```
$this->description = T_(
```
I think we finally did it!

lachlan-00 commented:
@Psy-Virus in spite 9f what I said . After all these new commits. NOW I'm ready for you to finalised :)

wagnered commented:
Hello Stebe, Manuel, lachan, kuzma fesenko and codacy bot,


Maybe this should be a separate, pinned issue, but I don't have access to the internet.

I don't have access to the internet, but I do have email.  I noticed that codacy is complaining about some security issues caused by some changes in this push request.

1.  "Detected usage of a non-validated input variable $_COOKIE "

2.  "Detected usage of a non-sanitized input variable $_COOKIE:

Please be careful and pay attention to codacy's issues,  especially those about security.

 Actually, if I remember correctly there is a section of the wiki that describes the Ampache design philosophy. Codacy is just trying to encourage us volunteer developers to follow our own philosophy.

I have been informed that the  Ampache entry in Wikipedia refers to the "Ampache coding pholosophy" and how important security is to it.  Let's mean it.

I would like to suggest that efforts to add new features be put on hold. It's time to take pride in developing Ampache. Do this by taking the time to modify the existing code so that  Ampache demonstrates the philosophy and the last issue a user will be concerned about is security. 

Regards,

Ernie D



-------- Original message --------
Subject: Re: [ampache/ampache] strings update (#1957)
From: lachlan <notifications@github.com>
To: ampache/ampache <ampache@noreply.github.com>
CC: Re: [ampache/ampache] strings update (#1957)

@Psy-Virus in spite 9f what I said . After all these new commits. NOW I'm ready for you to finalised :)

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub, or mute the thread.

lachlan-00 commented:
So good to hear from you!

I've put in a lot of variable sanitisation into the code with Core::get_* but nothing with the cookie variables so far. A lot of the 'new' issues I'm getting are because the line has changed making it fixed and new at the same time

The biggest open issue for security with me is the api which I'm almost done with.

Cleaning up codacy and scrutinizer issues has been a really fun process and I won't be stopping any time soon.

I fully agree that code quality is the most important feature Ampache should focus on. With this pull and API4 I'm done with features.

Again, really awesome to hear from you!

kuzi-moto commented:
Hey Ernie, long time no see! I totally agree that security should be taken seriously whenever working on Ampache.

That being said, I know barely any PHP/JavaScript to make me useful. So I mostly focus my efforts on smaller changes, most of which are cosmetic.

As I get more familiar with the code I will definitely try to tackle more of those issues.

wagnered commented:
Hey Kuzma,

We definitely appreciate any and all help you give us.

I learned a lot from [w3schools!](https://www.w3schools.com/php).  Do a search for "php filter".

PS: I tried to create the w3schools link from memory but I won't know for sure until I get my copy of this message. It might be backwards.

Probably a regex expression could be used to test the content of $_COOKIE[] using a php filter.

-------- Original message --------
Subject: Re: [ampache/ampache] strings update (#1957)
From: Kuzma Fesenko <notifications@github.com>
To: ampache/ampache <ampache@noreply.github.com>
CC: Re: [ampache/ampache] strings update (#1957)

Hey Ernie, long time no see! I totally agree that security should be taken seriously whenever working on Ampache.

That being said, I know barely any PHP/JavaScript to make me useful. So I mostly focus my efforts on smaller changes, most of which are cosmetic.

As I get more familiar with the code I will definitely try to tackle more of those issues.

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub, or mute the thread.

kuzi-moto commented:
Thanks for the tip! I know I've used w3schools before and has been a very useful resource. I'll definitely take a look and see if I can come up with a decent fix for it, unless someone beats me to it!

lachlan-00 commented:
@kuzi-moto it's probably filter_var you want.
https://www.php.net/manual/en/function.filter-var.php
https://www.php.net/manual/en/book.filter.php

As i said above though these aren't not new as in introduced; but 'fixed' and then reintroduced as 'new' because the line has changed.

lachlan-00 commented:
I've added Core::get_cookie for you to play with.

These get_* functions are just returning strings to if any variable is an array or requires an int just remember that these simple functions are for strings.

lachlan-00 commented:
Seeing as this has ballooned out to 600 commits and crazy amounts of comments i'm closing this to merge from #2062 instead.
